File: 295-find-median-from-data-stream/295-find-median-from-data-stream.cpp
class MedianFinder {
public:
    priority_queue<int> small;
    priority_queue<int,vector<int>, greater<int>>large;
    
    MedianFinder() {
        
    }
    
    void addNum(int num) {
        small.push(num);
        large.push(small.top());
        small.pop();
        if(large.size()>small.size()) {
            small.push(large.top());
            large.pop();
        }
    }
    
    double findMedian() {
        if(small.size()==large.size()){
            return (small.top() + large.top() )/2.0;
        }
        else{
            return small.top();
        }
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 295-find-median-from-data-stream/295-find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.maxHeap = []
        self.minHeap = []

    def addNum(self, num: int) -> None:
        heappush(self.maxHeap,-num)
        heappush(self.minHeap,-heappop(self.maxHeap))
        
        if len(self.minHeap)>len(self.maxHeap):
            heappush(self.maxHeap,-heappop(self.minHeap))

    def findMedian(self) -> float:
        if len(self.maxHeap)!=len(self.minHeap):
            return -self.maxHeap[0]
        else:
            return (-self.maxHeap[0]+self.minHeap[0])/2